src/coreclr/src/jit/lsra.cpp

```diff
@@ -2994,291 +2989,584 @@ regNumber LinearScan::tryAllocateFreeReg(Interval* currentInterval, RefPosition*
-        {
-            score |= VALUE_AVAILABLE;
-        }
+    // At this point we have determined all the candidates that CAN be considered,
+    // and we'll begin to apply the heuristics in order.
+    registerSelector selector;
+    selector.candidates = candidates;
+    selector.score      = 0;
+#ifdef TARGET_ARM
+    selector.regType = regType;
+#endif // TARGET_ARM
 
-        // If the nextPhysRefLocation is a fixedRef for the rangeEndRefPosition, increment it so that
-        // we don't think it isn't covering the live range.
-        // This doesn't handle the case where earlier RefPositions for this Interval are also
-        // FixedRefs of this regNum, but at least those are only interesting in the case where those
-        // are "local last uses" of the Interval - otherwise the liveRange would interfere with the reg.
-        if (nextPhysRefLocation == rangeEndLocation && rangeEndRefPosition->isFixedRefOfReg(regNum))
-        {
-            INDEBUG(dumpLsraAllocationEvent(LSRA_EVENT_INCREMENT_RANGE_END, currentInterval));
-            nextPhysRefLocation++;
-        }
+    freeCandidates = getFreeCandidates(candidates, regType);
 
-        if ((candidateBit & preferences) != RBM_NONE)
-        {
-            score |= OWN_PREFERENCE;
-            if (nextPhysRefLocation > rangeEndLocation)
-            {
-                score |= COVERS;
-            }
-        }
-        if ((candidateBit & relatedPreferences) != RBM_NONE)
+    // Apply the FREE heuristic.
+    if (freeCandidates == RBM_NONE)
+    {
+        // We won't spill if this refPosition is not an actual ref.
+        if (!refPosition->IsActualRef())
         {
-            score |= RELATED_PREFERENCE;
-            if (nextPhysRefLocation > relatedLastLocation)
-            {
-                score |= COVERS_RELATED;
-            }
+            currentInterval->assignedReg = nullptr;
+            return REG_NA;
         }
+    }
+    else if (!found)
+    {
+        found = selector.applySelection(FREE, freeCandidates);
+    }
 
-        // If we had a fixed-reg def of a reg that will be killed before the use, prefer it to any other registers
-        // with the same score.  (Note that we haven't changed the original registerAssignment on the RefPosition).
-        // Overload the RELATED_PREFERENCE value.
-        else if (candidateBit == refPosition->registerAssignment)
+    // Apply the VALUE_AVAILABLE (matching constant) heuristic. Only applies if we have freeCandidates.
+    // Note that we always need to define the 'matchingConstants' set.
+    if (freeCandidates != RBM_NONE)
+    {
+        if (currentInterval->isConstant && RefTypeIsDef(refPosition->refType))
         {
-            score |= RELATED_PREFERENCE;
+            matchingConstants = getMatchingConstants(selector.candidates, currentInterval, refPosition);
+            found             = selector.applySelection(VALUE_AVAILABLE, matchingConstants);
```

Comment:
Yes, good catch. If `found` was already true, we still want to check to see if it's a matching constant (so that we know to mark it for reuse). But as you say, we don't want to then reset `found` to false. It will still produce the same result, as we won't be adding back any candidates, but it will just do unnecessary work.

Reply:
Alright, changed to `found |= `.